source/common/api/posix/os_sys_calls_impl.cc

```diff
@@ -110,6 +110,43 @@ bool OsSysCallsImpl::supportsUdpGso() const {
 #endif
 }
 
+bool OsSysCallsImpl::supportsIpTransparent() const {
+#if !defined(__linux__)
+  return false;
+#else
+  // The linux kernel supports IP_TRANSPARENT by following patch(starting from v2.6.28) :
+  // https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/net/ipv4/ip_sockglue.c?id=f5715aea4564f233767ea1d944b2637a5fd7cd2e
+  //
+  // The linux kernel supports IPV6_TRANSPARENT by following patch(starting from v2.6.37) :
+  // https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/net/ipv6/ipv6_sockglue.c?id=6c46862280c5f55eda7750391bc65cd7e08c7535
+  //
+  // So, almost recent linux kernel supports both IP_TRANSPARENT and IPV6_TRANSPARENT options.
```

Comment:
Yes. That is correct. So, I check the both options and throw an exception if both of them are not supported during initialization phase.

Reply:
My point is that we don't even need to check that, given that it's supported for more than 10 years.

Reply:
Because even if it is supported for more then 10 years, it needs the NET_ADMIN capability that is set by root privilege. Please refer to following man page. https://man7.org/linux/man-pages/man7/ip.7.html
So, it should be checked on runtime if the envoy process has proper a capability.
In that case, check if have only NET_ADMIN capability is enough but set the actual socket option is more general checking method I think.

Reply:
If we should consider ipv4 only enabled linux kernel, I think check the listener's ip address during `UdpProxyFilter` creation.
I added some logs on the constructor of `UdpProxyFilter`, I could read it.

`UdpProxyFilter::UdpProxyFilter(Network::UdpReadFilterCallbacks& callbacks,
                               const UdpProxyFilterConfigSharedPtr& config)
    : UdpListenerReadFilter(callbacks), config_(config),
      cluster_update_callbacks_(
          config->clusterManager().addThreadLocalClusterUpdateCallbacks(*this)) {
  Upstream::ThreadLocalCluster* cluster = config->clusterManager().get(config->cluster());
  if (cluster != nullptr) {
    onClusterAddOrUpdate(*cluster);
  }
  ENVOY_LOG(info, "lisenter ip: {}", callbacks.udpListener().localAddress()->asStringView());
}`

>[2020-08-21 13:05:39.332][24681][info][main] [source/server/server.cc:644] all clusters initialized. initializing init manager
[2020-08-21 13:05:39.332][24681][info][config] [source/server/listener_manager_impl.cc:884] all dependencies initialized. starting workers
[2020-08-21 13:05:39.333][24681][info][main] [source/server/server.cc:663] starting main dispatch loop
[2020-08-21 13:05:39.340][24687][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324
[2020-08-21 13:05:39.341][24690][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324
[2020-08-21 13:05:39.344][24700][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324
[2020-08-21 13:05:39.344][24691][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324
[2020-08-21 13:05:39.344][24688][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324
[2020-08-21 13:05:39.344][24698][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324
[2020-08-21 13:05:39.346][24696][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324
[2020-08-21 13:05:39.346][24694][info][filter] [source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:21] lisenter ip: 172.17.0.1:62324

Please give your opinion and let me know if it is proper approach.